packages/webapp/src/components/Table/Table.tsx

```diff
@@ -0,0 +1,34 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+import React from 'react';
+import ReactTable from './index';
+import MuiTable from './v2';
+import { TableType, KindComponentKVP, ReactTableProps, MuiTableProps } from './types';
+
+const tableComponents: KindComponentKVP = {
+  [TableType.V1]: (props: ReactTableProps) => <ReactTable {...props} />,
+  [TableType.V2]: (props: MuiTableProps) => <MuiTable {...props} />,
+};
+
+/**
+ * A component that selects between available Table styles.
+ * See packages/webapp/src/stories/Table/Table.stories.jsx for examples.
+ */
+// TODO: export default function Table({ kind, ...props } : {kind:TableType, props: any | ReacttableProps | MuiTableProps}) {
+export default function Table({ kind, ...props }: any) {
```

Comment:
Cheating using the any type here couldn't get props to reduce scope here.

Reply:
This is kind of a gnarly component to be typing with the key value pair mapping and the double wrapping!!

If you don't mind doing the key value pair mapping right in the component itself (instead of with an intermediary constant), you could write a discriminated union for the props and get the type checking on the component when you call it based on what you provide for `kind`):

```ts
type ReactTable = ReactTableProps & { kind: TableType.V1 };
type MuiTable = MuiTableProps & { kind: TableType.V2 };

type TableProps = ReactTable | MuiTable;

export default function Table({ kind, ...props }: TableProps) {
  switch (kind) {
    case TableType.V1:
      return <ReactTable {...(props as ReactTableProps)} />;
    case TableType.V2:
      return <MuiTable {...(props as MuiTableProps)} />;
  }
}

```

Would that work for you? I don't know how to keep `KindComponentKVP` as a separate object and still keep the narrowing.

Reply:
Nice thank you! I am still only like 1 or 2 hrs in on my TS course. 

Okay -- I think I really like this pattern and intend to use it moving forward since it is quite effective for passing along the typescript typings. There only seems to *maybe* be a performance hit for large numbers of switch statements vs lookup table (which I don't think would happen in our cases) (https://stackoverflow.com/questions/37730199/switch-vs-object-lookup-performance-since-jsperf-is-down, https://stackoverflow.com/questions/11617091/in-a-switch-vs-dictionary-for-a-value-of-func-which-is-faster-and-why) 

Thanks for sharing the keyword the [discriminated union docs ](https://basarat.gitbook.io/typescript/type-system/discriminated-unions) are really helpful, I also looked into [Strategy Patterns](https://dev.to/carlillo/design-patterns---strategy-pattern-in-javascript-2hg3) this morning to see if there was anything recommended or popular -- but didn't come up with much.